simulations/csh-benchmarks/R/eval-functions.R

```diff
@@ -0,0 +1,231 @@
+
+  ## extract the true variance parameter values
+  hom_vars <- diag(true_covar_mat_ls[[1]])
+  het_vars <- diag(true_covar_mat_ls[[2]])
+
+  ## extract the true correlation parameter value
+  hom_corr <- true_covar_mat_ls[[1]][1, 2] / sqrt(hom_vars[1] * hom_vars[2])
+  het_corr <- true_covar_mat_ls[[2]][1, 2] / sqrt(het_vars[1] * het_vars[2])
+
+  ## compute the squared error losses
+  fit_results %>%
+    dplyr::mutate(
+      sq_err_loss = purrr::map2_dfr(
+        fit, .dgp_name,
+        function(f, dgp_name) {
+          ## extract the estimated variances and correlation
+          est_covar_mat <- get_covar_mat(f)
+          vars_est <- diag(est_covar_mat)
+          corr_est <- est_covar_mat[1, 2] / sqrt(vars_est[1] * vars_est[2])
+
+          ## assemble the true parameter vector
+          if (dgp_name == "hom_rct")
+            true_params <- c(hom_vars, hom_corr)
+          else
+            true_params <- c(het_vars, het_corr)
+
+          ## compute the squared error loss
+          loss <- (c(vars_est, corr_est) - true_params)^2
+          names(loss) <- c(paste0("var_t", seq_len(10)), "corr")
+          return(loss)
+
+        }
+      )
+    ) %>%
+    dplyr::select(-fit) %>%
+    tidyr::unnest(cols = sq_err_loss)
+
+}
+
+#' Bias of heterogeneous compound symmetry matrix parameters
+#'
+#' @description This function computes the bias of the individual
+#'   compound symmetry matrix parameters. These parameters are the variance
+#'   terms at each time point, as well as the correlation coefficient.
+#'
+#' @param fit_results A tibble of simulation replicates generated by simChef. It
+#'   contains the fitted mixed models for all model implementations.
+#'  @param true_covar_mat_ls A named list containing the true repeated measures
+#'   covariance matrices associated with each data-generating process. The
+#'   matrices' names must match those of the data-generating processes.
+#'
+#' @return A tibble containing the empirical bias for each parameter of the
+#'   intra-observation covariance matrix, stratified by data-generating process,
+#'   method and sample size.
+csh_param_bias_fun <- function(fit_results, true_covar_mat_ls) {
+
+  ## extract the true variance parameter values
+  hom_vars <- diag(true_covar_mat_ls[[1]])
+  het_vars <- diag(true_covar_mat_ls[[2]])
+
+  ## extract the true correlation parameter value
+  hom_corr <- true_covar_mat_ls[[1]][1, 2] / sqrt(hom_vars[1] * hom_vars[2])
+  het_corr <- true_covar_mat_ls[[2]][1, 2] / sqrt(het_vars[1] * het_vars[2])
+
+  ## compute the biases
+  strata_vars <- c(".dgp_name", ".method_name", "num_part")
+  fit_results %>%
+    dplyr::mutate(
+      accuracy = purrr::map2_dfr(
+        fit, .dgp_name,
+        function(f, dgp_name) {
+          ## extract the estimated variances and correlation
+          est_covar_mat <- get_covar_mat(f)
+          vars_est <- diag(est_covar_mat)
+          corr_est <- est_covar_mat[1, 2] / sqrt(vars_est[1] * vars_est[2])
```

Comment:
hm not sure but why is this just 2x2 here? I thought we have 10 variances in the heterogeneous case? plus a correlation

Reply:
Line 196 extracts all variances from the covariance matrix estimate. Line 197 is used to compute the correlation estimate based on the CSH model. I'm using the estimated correlation of the first and second time points here, but any other pair of time points could be used. I'll add comments to explain the logic.